pilot/pkg/serviceregistry/serviceentry/controller.go

```diff
@@ -917,42 +917,52 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 			hash.Reset()
 		}
 	}
-	// i is everything from 240.240.0.(j) to 240.240.255.(j)
-	// j is everything from 240.240.(i).1 to 240.240.(i).254
-	// we can capture this in one integer variable.
-	// given X, we can compute i by X/255, and j is X%255
-	// To avoid allocating 240.240.(i).255, if X % 255 is 0, increment X.
-	// For example, when X=510, the resulting IP would be 240.240.2.0 (invalid)
-	// So we bump X to 511, so that the resulting IP is 240.240.2.1
+
 	x := 0
-	hnMap := make(map[string]octetPair)
+	y := 0
 	allocated := 0
+	hnMap := make(map[string]octetPair)
 	for _, svc := range hashedServices {
 		if svc == nil {
 			// There is no service in the slot. Just increment x and move forward.
 			x++
-			if x%255 == 0 {
-				x++
-			}
 			continue
 		}
 		n := makeServiceKey(svc)
 		if v, ok := hnMap[n]; ok {
 			log.Debugf("Reuse IP for domain %s", n)
 			setAutoAllocatedIPs(svc, v)
 		} else {
-			x++
-			if x%255 == 0 {
+			allocated++
+			var thirdOctect, fourthOctect int
+			if x/255 < 255 {
+				// To avoid allocating 240.240.(i).255, if X % 255 is 0, increment X.
+				// For example, when X=510, the resulting IP would be 240.240.2.0 (invalid)
+				// So we bump X to 511, so that the resulting IP is 240.240.2.1
 				x++
+				if x%255 == 0 {
+					x++
+				}
 			}
-			if allocated >= maxIPs {
-				log.Errorf("out of IPs to allocate for service entries. x:= %d, maxips:= %d", x, maxIPs)
-				return services
+			thirdOctect = x / 255
+			// When we reach thirdOctect 255, we need to just increment the fourthOctect
+			// keeping the thirdOctect as 255.
+			if thirdOctect >= 255 {
+				y++
+				thirdOctect = 255
+				fourthOctect = y
```

Comment:
won't this conflict with a pre allocated svc?
IIUC, this means the svc auto allocated is out of the range supported, actually exceed 255 * 255

Reply:
it is not. it just means third octet is full and we still have room for [1-255] in the fourth octet

Reply:
maybe i am misreading it: how could x / 255 = 256, because MAXIP is 255*255

Reply:
If x/255 ==255 || x/255 = 256
   y++

Then i am afraid y could exceed 255

Reply:
> maybe i am misreading it: how could x / 255 = 256, because MAXIP is 255*255

because `x` is different from allocated. We increment x to skip when svc is not available in the slot. 

Please see the updated test case - it captures the behaviour. you can run it on master without the fix and see the difference

Reply:
Then for both case `x/255 ==255` and  `x/255 = 256`, the fourth octet y++？

woun't y > 255?

My doubt is we donot need to do it for case x/255 ==255

Reply:
for 240.240.x.x, we can at most allocate 256(0-255) * 254 (1-254) = 65024 ips, but here in istio we set MAXIP= 255 * 255 = 65025.

So in theory,  we must have a conflict if the service number exceed a value

Reply:
`if allocated > maxIPs`  - Would n't this condition help ? Are you suggesting we set MAXIPS to 65024?

Reply:
Yeah, i think we have to